frontend/src/containers/EditSchedule/utils/index.js

```diff
@@ -0,0 +1,25 @@
+import { getCourses } from "services/api"
+
+
+
+export const generateScheduledCourseListFromSchedule = async (majorId, schedule) => {
+    const { data: courses } = await getCourses(majorId);
+    let result = [];
+    let added = {};
+
+    for (let i = 0; i < courses.courses.length; i++) {
+        for (let j = 0; j < courses.courses[i].classes.length; j++) {
+            for (let k = 0; k < schedule.schedule_items.length; k++) {
+                if (schedule.schedule_items[k].name === courses.courses[i].classes[j].name) {
+                    if (!added[courses.courses[i].classes[j].name]) {
+                        result.push({ ...courses.courses[i].classes[j], credit: courses.courses[i].credit, parentName: courses.courses[i].name, term: courses.courses[i].term, name: courses.courses[i].classes[j].name });
+                        added[courses.courses[i].classes[j].name] = 1;
+                    }
+                }
+            }
+        }
+    }
```

Comment:
Desperately in need of refactoring. This kind of code will not get into production 😂 @leleonardo81 @laymonage

Reply:
We have a feature that lets you define your own agenda to be added to the schedule. So, `schedule_items` may contain items that do not exist in the current `courses`.

I'd suggest that:

- We make a copy of `schedule_items` e.g. `savedSchedule`; then
- in addition to `push`ing the adapted `schedule` to the `result` array, we should also remove the matching schedule from the `savedSchedule` array; then
- we iterate through the remaining items in the `savedSchedule` (these are items that don't exist in the current `courses`: they may be custom agendas or courses from the previous terms), adapt them, and push them to the `results` array.

Reply:
Ah yes, didn't really think about the custom agendas when implementing this. Thanks for the proposed solutions, will implement this asap.